### PR TITLE
handle UDQ value for WINJFCNC

### PR DESCRIPTION
**Pull request by GitPaean**
restarting related is missing for now.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 please

**Comment by GitPaean** [comment]
The modeling of filtration during injection is something we only have in OPM-flow. Not sure whether we should have it in the RST support. @bska can comment on this.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 please

**Comment by GitPaean** [comment]
Will investigate the jenkins failures.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 please

**Comment by GitPaean** [comment]
With the small revision from the opm-simulators side when coming to arithmetic floating point calculation, the simulation results are identical with the master branch. 

The only thing changes is the changes of the RESTART keywords due to the use the UDQ in the opm-tests, 
```
Unified restart files, sequence  1

not same keywords in Restart, sequence 1

  DOUBHEAD |   DOUBHEAD
           |       DUDF
    ENDSOL |     ENDSOL
      ICON |       ICON
      IGRP |       IGRP
  INTEHEAD |   INTEHEAD
           |       IUDQ
      IWEL |       IWEL
      IWLS |       IWLS
  LOGIHEAD |   LOGIHEAD
      PPCW |       PPCW
  PRESSURE |   PRESSURE
        RS |         RS
        RV |         RV
      SCON |       SCON
    SEQNUM |     SEQNUM
      SGAS |       SGAS
      SGRP |       SGRP
  STARTSOL |   STARTSOL
      SWAT |       SWAT
      SWEL |       SWEL
  THRESHPR |   THRESHPR
      XCON |       XCON
      XGRP |       XGRP
      XWEL |       XWEL
      ZGRP |       ZGRP
           |       ZUDL
           |       ZUDN
      ZWEL |       ZWEL
      ZWLS |       ZWLS

Program threw an exception: [/var/lib/jenkins/post-builder/workspace/opm-common-PR-builder/test_util/EclRegressionTest.cpp:275] 
Keywords not identical in Restart, sequence 1
```

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 ignore_extra please

**Comment by bska** [comment]
> ```
> Value index                 = 262
> (first value, second value) = (0, 1)
> ```

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp [hunk 1]
```diff
@@ -2002,9 +2002,8 @@ Well{0} entered with 'FIELD' parent group:
             const auto well_names = this->wellNames(wellNamePattern, handlerContext);
             for (const auto& well_name: well_names) {
                 auto well = this->snapshots.back().wells(well_name);
-                const auto filter_conc = record.getItem<ParserKeywords::WINJFCNC::VOL_CONCENTRATION>().get<double>(0);
-                // the unit is ppm_vol
-                well.setFilterConc(filter_conc/1.e6);
+                const auto filter_conc = record.getItem<ParserKeywords::WINJFCNC::VOL_CONCENTRATION>().get<UDAValue>(0);
+                well.setFilterConc(filter_conc );
```
Comment (bska) [inline src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2006]: What happened to the factor of `1.0e6`?  I don't see this in the unit handling.

**Review by bska — COMMENTED** [review]
I don't quite understand the unit handling here.  The JSON keyword model says that the concentration [does not have any units](https://github.com/OPM/opm-common/blob/e51c5cb4dd11ec12cc799670062bc808ecdb8313/src/opm/input/eclipse/share/keywords/900_OPM/W/WINJFCNC#L14), but the original code treats the value as a parts per million volume concentration.

Which is the expected behaviour?

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp (hunk 1 above)
Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2006]: I moved to scale in the simulator side, https://github.com/OPM/opm-simulators/pull/4763/files#diff-27ec506a141c179a9728b3453b16c0df44e43b734900be39b2a4709fc0ee4316L71 . 

```c++
filtration_particle_volume_[perf] += water_rates * (conc * 1.e-6) * dt; // conc is with ppm_vol unit
```

Basically I really do not know how we should handle the PPM volume unit properly due to the limited view of the all the components. 

Will a dimension `1.e-6` in the json file do anything better in this direction. 
```
    {
      "name": "VOL_CONCENTRATION",
      "value_type": "DOUBLE",
      "dimension": "1e-6",
      "default" : 0
    }
```
And also, should I use the `Dimension` in the class `UDAValue` to make things better presented?  Suggestions on these will be much appreciated.

Comment (bska) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2006]: > Will a dimension `1.e-6` in the json file do anything better in this direction.

No, not really.  I would actually hope you'd get a build failure if you tried to write that in the JSON file.  The only permissible `"dimension"` strings are supposed to be (combinations of) the dimensions explicitly registered/added in [`UnitSystem.cpp`](https://github.com/OPM/opm-common/blob/0895983ac1ccc6dc6aa2bde88a047419327d97e4/src/opm/input/eclipse/Units/UnitSystem.cpp#L1181-L1207).  Maybe we could
```C++
addDimension("PPM", 1.0e-6);
```
or something to that effect in all unit conventions and then switch the JSON-file to using `"dimension" : "PPM"`?

> And also, should I use the `Dimension` in the class `UDAValue` to make things better presented?

If we do `addDimension("PPM", 1.0e-6)`, then we should definitely use `UDAValue::get_dim().convertRawToSi()` to compute the concentration value.  That is more or less exactly how we compute target/limit values in `WCON*` and `GCON*` that are given by UDAs.  As an example, please have a look at [`OPM::UDA::eval_well_uda()`](https://github.com/OPM/opm-common/blob/0895983ac1ccc6dc6aa2bde88a047419327d97e4/src/opm/input/eclipse/Schedule/eval_uda.cpp#L47).

Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2006]: Thanks. I will try something along this line and update the PR.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 ignore_extra please

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp (hunk 1 above)
Comment (GitPaean) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2006]: Now I use the newly introduced `PPM` unit. It works well.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 ignore_extra please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 ignore_extra please

**Comment by bska** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 update_data please

**Comment by bska** [comment]
> jenkins build this [...] update_data 

Hm, that ended up producing a change to [NORNE_ATW2013.INIT](https://github.com/OPM/opm-tests/pull/1022) which is highly unexpected.  I don't think the reason is this particular PR, rather I suspect it's fallout from some other work recently, but I'll hold off on merging this until we've discovered the underlying problem.

**Comment by atgeirr** [comment]
> Hm, that ended up producing a change to [NORNE_ATW2013.INIT](https://github.com/OPM/opm-tests/pull/1022) which is highly unexpected. I don't think the reason is this particular PR, rather I suspect it's fallout from some other work recently, but I'll hold off on merging this until we've discovered the underlying problem.

I agree, this should be high priority! If I understand the situation correctly, the changes in the .INIT file were small enough to still let the test pass, so it is probably not a critical flaw. Nonetheless, changes where we do not expect them are always suspicious.

**Comment by akva2** [comment]
i think this is just an ordering thing. I reordered a table in alpha order https://github.com/OPM/opm-common/commit/191364d2f3723037c9bb1afdb98178c4043b7d96, and they thus ended up in different order in the file. the test pass fine since the tables are exactly the same, however the file has changed. so whatever update we do now it will include the .INIT file.

**Comment by bska** [comment]
> i think this is just an ordering thing. I reordered a table in alpha order https://github.com/OPM/opm-common/commit/191364d2f3723037c9bb1afdb98178c4043b7d96, and they thus end up in different order in the file.

Right.  I kind of wish we didn't do that though since there was a reason for the original order.  In the grand scheme of things I don't think it actually matters though since the numerical values are the same and no tool/workflow uses the `.INIT` file other than for visualisation purposes.

Just for reference, below is the difference in [`arraylist`](https://github.com/OPM/opm-common/blob/99e3608bf36deac109b1ecb3f42fe0ac543f6140/test_util/arraylist.cpp) output between the original (i.e., prior to 191364d2f3723037c9bb1afdb98178c4043b7d96) version (`-`) and current master (`+`) version of `NORNE_ATW2013.INIT`.

```diff
--- /dev/fd/63	2023-08-18 09:48:53.760267195 +0200
+++ /dev/fd/62	2023-08-18 09:48:53.760267195 +0200
@@ -6,11 +6,11 @@
 DX              44431   REAL
 DY              44431   REAL
 DZ              44431   REAL
+NTG             44431   REAL
 PORO            44431   REAL
 PERMX           44431   REAL
 PERMY           44431   REAL
 PERMZ           44431   REAL
-NTG             44431   REAL
 TRANX           44431   REAL
 TRANY           44431   REAL
 TRANZ           44431   REAL
```

**Comment by atgeirr** [comment]
I think the old order should be restored (a very small change), maybe with a brief comment in the code that the order is chosen to be similar to Eclipse. Then this PR should not generate an INIT update.

**Comment by akva2** [comment]
https://github.com/OPM/opm-common/pull/3650

**Comment by bska** [comment]
> #3650

Thanks!  @GitPaean: We've merged the above PR now and I **think** that will be sufficient to ensure that when we update the reference solutions for this PR there will be no unexpected changes.  Just to be on the safe side, however, would you mind rebasing this PR onto the current master sources please?

**Comment by GitPaean** [comment]
> Just to be on the safe side, however, would you mind rebasing this PR onto the current master sources please?

Sure. I have done it. Let us how the testing goes.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 update_data please

**Comment by GitPaean** [comment]
since we ware also changing the DATA files in opm-tests. I will manually put the update_data PR into  https://github.com/OPM/opm-tests/pull/1025 , so Jenkins can verify.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 please

**Comment by bska** [comment]
> please help to look at the Jenkins testing, to make sure nothing fishy there

It looks good to me.  No unexpected changes there as far as I can tell.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=4763 opm-tests=1016 please

**Review by bska — APPROVED** [review]
Thanks a lot for your patience and for the updates.  This looks good to me now and the revised reference solutions have been installed on the CI system.  I'll merge this into the master branch now.

**Pull request merged**